### ISSUE TITLE
NTI: Invalid warning when using optional properties

### ISSUE THREAD

**Issue by vobruba-martin** [google/closure-compiler#1421]
I've update from `62ca536` to `825a571` and now I'm getting this warning:

```javascript

/**
 * @typedef {{
 * 	x : (undefined|string)
 * }}
 */
var MyOptions;


/**
 * @param {MyOptions} opts
 */
function takeOptions(opts) {
	/**
	 * @type {string}
	 */
	var x = opts.x || 'some string';
};
```
> WARNING - The right side in the assignment is not a subtype of the left side.
> left side  : string
> right side : string|undefined
> 
> 	var x = opts.x || 'some string';
> 	        ^
> 
> 0 error(s), 1 warning(s), 100.0% typed

Same is with `@record`:

```javascript

/**
 * @record
 */
function MyOptions() {};

/**
 * @type {(undefined|string)}
 */
MyOptions.prototype.x;
```

**Comment by MatrixFrog** [comment]
https://closure-compiler-debugger.appspot.com/#input0%3D%252F**%250A%2520*%2520%2540typedef%2520%257B%257B%250A%2520*%2520%2520x%2520%253A%2520(undefined%257Cstring)%250A%2520*%2520%257D%257D%250A%2520*%252F%250Avar%2520MyOptions%253B%250A%250A%250A%252F**%250A%2520*%2520%2540param%2520%257BMyOptions%257D%2520opts%250A%2520*%252F%250Afunction%2520takeOptions(opts)%2520%257B%250A%2520%2520%2520%2520%252F**%250A%2520%2520%2520%2520%2520*%2520%2540type%2520%257Bstring%257D%250A%2520%2520%2520%2520%2520*%252F%250A%2520%2520%2520%2520var%2520x%2520%253D%2520opts.x%2520%257C%257C%2520'some%2520string'%253B%250A%257D%253B%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

**Comment by dimvar** [comment]
sorry for the regression. Working on a fix.
